Add simulcast sending support

kc5nra opened the pull request:
Adds support for specifying the RID when writing packets to str0m as well as for serialization of the VLA extended header.

There are a couple points that need to be resolved:
* Carrying the active/inactive state of streams from simulcast specification (adding the media) to when the SDP is created
* A couple of the VLA tests I wasn't sure about. Specifically the one that I did convert add the bidirectional serialization to (it says 3 active streams but it has no active spatial layers, which would end up being condensed to the shared bitmask)

algesten commented on src/util/bit_writer.rs:
```diff
@@ -0,0 +1,439 @@
+    buf: &'a mut [u8],
+    index: usize,
+    bit_index: u8,
+}
```
We get an unfortunate asymmetry between reading and writing VLA where the reading does not have this kind of helper.

Spontaneously it looks a bit over engineered to require ~400 lines of source file to write some bits in a (for the time being) fringe RTP header extension, in a direction that is not a super common use case (since I'm assuming receiving simulcast is more common than sending).

I'm not saying we definitely shouldn't commit this, but I think we need to reason about the future of it and whether it solves a problem hard enough to not be a few helper methods.

kc5nra replied:
Np! I was kinda playing with compile time assertions to prevent runtime checks. My original version was in situ so I'll just swap it out

kc5nra replied:
Done

algesten commented on src/media/mod.rs:
```diff
@@ -525,6 +525,27 @@ impl Media {
+                })
+            } else {
+                None
+            },
```
Refactor this so that you can write `a.simulcast.map(|s| s.into_sdp()` and move the actual code body elsewhere.

kc5nra replied:
Done

algesten reviewed (commented):
Thanks!

This looks good overall. I left some comments.

In a later PR we should also toy with the idea of creating a new type `MidRid` that looks something like:

```rs
pub(crate) struct MidRid {
   pub(crate) mid: Mid,
   pub(crate) rid: Option<Rid>,
}
```

to see if it would tidy up code to avoid passing around tuples in many places.

I'll leave it to @k0nserv to say whether it's correct to run the pacer per-mid/rid combo. I think it is, but better check with the expert.

thomaseizinger commented on src/rtp/vla.rs:
```diff
@@ -211,10 +213,137 @@ impl VideoLayersAllocation {
+
+                    sl_bm
+                })
+                .collect::<Vec<_>>();
```
`smallvec` is pretty popular for this sort of stuff.

kc5nra replied:
Removed most if not all vecs from this code

kc5nra replied:
We can figure out a better approach (fixed arrays or SmallVec-like solution) later. These changes at least don't add additional ones (I was reluctant to add another dependency before more consensus occurred)

kc5nra commented on src/streams/send.rs:
```diff
@@ -668,7 +668,7 @@ impl StreamTx {
         // The packet remains in the head of the send queue until we
         // finish poll_packet, at which point we move it to the cache.
-        let pkt = self.send_queue.peek()?;
+        let pkt: &mut RtpPacket = self.send_queue.peek()?;
```
Nope, from some earlier debugging. Will fix

kc5nra replied:
Done

kc5nra commented on src/rtp/vla.rs:
```diff
@@ -261,6 +389,21 @@ fn parse_leb_u63(bytes: &[u8]) -> (u64, &[u8]) {
 }
 
+/// Encodes leb128
+pub fn encode_leb_u64(mut value: u64) -> Vec<u8> {
```
Ya - I just made no attempt to prevent a full 64bit value being encoded so just named it.. I also noticed the asymmetry but it wouldn't actually being u63 for the encode side

kc5nra replied:
Ya - I just made no attempt to prevent a full 64bit value being encoded so just named it.. I also noticed the asymmetry but it wouldn't actually being u63 for the encode side

algesten replied:
Cool. Maybe it doesn't matter because no one is encoding full 64 bits with it anyway?

kc5nra replied:
Renamed and made to use underlying buf instead of vec

kc5nra commented:
I also thought that but just wanted your opinion first - will quickly refactor it

kc5nra commented:
I figured since we are adding new code we could switch (or not!) to midrid, and possibly move the stats in another commit if we like it

kc5nra commented on src/change/sdp.rs:
```diff
@@ -550,6 +555,7 @@ pub(crate) struct AddMedia {
     pub kind: MediaKind,
     pub dir: Direction,
     pub ssrcs: Vec<(Ssrc, Option<Ssrc>)>,
+    pub simulcast: Option<crate::media::Simulcast>,
```
Yeah, this was an API breaking change

algesten commented:
I turned `MidRid` into a tuple because it seemed more natural and makes for tidier log messages for the `fmt::Debug`.

I also implemented in many more places. I think I'll break the whole `MidRid` refactoring out to a separate PR to make this PR focused on simulcast.

algesten commented:
Heads up: I force pushed a rebase off latest main.

algesten commented:
MidRid refactoring in #604

thomaseizinger commented:
Didn't mean to elicit a review request with my drive-by comments :D
I am not familiar with these parts of the codebase so I am going to dismiss the review request.

algesten commented:
> Didn't mean to elicit a review request with my drive-by comments :D I am not familiar with these parts of the codebase so I am going to dismiss the review request.

All code must be scrutinized! :D

thomaseizinger commented:
> > Didn't mean to elicit a review request with my drive-by comments :D I am not familiar with these parts of the codebase so I am going to dismiss the review request.
> 
> All code must be scrutinized! :D

Say no more!

k0nserv commented:
> I'll leave it to @k0nserv to say whether it's correct to run the pacer per-mid/rid combo. I think it is, but better check with the expert.

Yes it's fine. The pacer only uses mid as a lookup key, it could equally use SSRC or this new combo of `(Mid, Option<Rid>)`.

algesten commented on src/change/sdp.rs:
```diff
@@ -223,14 +223,15 @@ impl<'a> SdpApi<'a> {
         dir: Direction,
         stream_id: Option<String>,
         track_id: Option<String>,
+        simulcast: Option<crate::media::Simulcast>,
```
Yes. I will do a follow-up PR turning this into a builder pattern.

algesten commented:
@kc5nra I rebased/squashed commits into two commits. One for the simulcast sending support and one for the VLA serialization. This is how I want this PR to land in main. If we do more work, we can fixup to those two.

There are two places I think we need to address before this PR is complete:

* https://github.com/algesten/str0m/blob/main/src/change/sdp.rs#L898
* https://github.com/algesten/str0m/blob/main/src/change/sdp.rs#L1305

algesten commented:
@kc5nra I will now leave this PR in peace (no more force pushing from my side) to let you catch up.

kc5nra commented:
> @kc5nra I rebased/squashed commits into two commits. One for the simulcast sending support and one for the VLA serialization. This is how I want this PR to land in main. If we do more work, we can fixup to those two.
> 
> There are two places I think we need to address before this PR is complete:
> 
> * https://github.com/algesten/str0m/blob/main/src/change/sdp.rs#L898
> * https://github.com/algesten/str0m/blob/main/src/change/sdp.rs#L1305

For https://github.com/algesten/str0m/blob/main/src/change/sdp.rs#L898, you mean something different than this? Though I agree we currently hand wave the active state

https://github.com/algesten/str0m/pull/603/commits/ad2e78c798109ea0dbd4652ea96fba9330a600a9#diff-5fd5a802394cab016425ddca33d909c07672077403390ccafa359c7d55dcbed0R903-R907

The latter isn't totally necessary for simulcast + VLA (late binding rids to ssrc), but it's 4AM and I'm probably not thinking of all the other permutations

algesten commented:
> > @kc5nra I rebased/squashed commits into two commits. One for the simulcast sending support and one for the VLA serialization. This is how I want this PR to land in main. If we do more work, we can fixup to those two.
> > There are two places I think we need to address before this PR is complete:
> > 
> > * https://github.com/algesten/str0m/blob/main/src/change/sdp.rs#L898
> > * https://github.com/algesten/str0m/blob/main/src/change/sdp.rs#L1305
> 
> For https://github.com/algesten/str0m/blob/main/src/change/sdp.rs#L898, you mean something different than this? Though I agree we currently hand wave the active state



> [ad2e78c#diff-5fd5a802394cab016425ddca33d909c07672077403390ccafa359c7d55dcbed0R903-R907](https://github.com/algesten/str0m/commit/ad2e78c798109ea0dbd4652ea96fba9330a600a9#diff-5fd5a802394cab016425ddca33d909c07672077403390ccafa359c7d55dcbed0R903-R907)
> 
> The latter isn't totally necessary for simulcast + VLA (late binding rids to ssrc), but it's 4AM and I'm probably not thinking of all the other permutations

Go to bed!

When we negotiate simulcast with Firefox, we get these rows:

<img width="570" alt="image" src="https://github.com/user-attachments/assets/5affe0d7-1e11-40cb-95e8-da7755b0e4cf" />

There are two (three?) mechanisms at play, and we need to support them all, to some extent.

1. Old school, pre-communicated SSRC in the SDP. The last three `group:FID` are declared in the order of the `a=simulcast`, i.e. l, m, h in this case. However for Firefox this is incorrect, so we use mechanism 2 https://github.com/algesten/str0m/blob/21355864916ffc373f05aa3c1e90582719f41e17/src/change/sdp.rs#L1113-L1116
2. Dynamic SSRC mapping using RTP header extensions. https://github.com/algesten/str0m/blob/main/src/session.rs#L309-L311
3. (I'm unsure here) VLA?

For incoming simulcast, we should ideally handle 1, but we're currently relying only on 2 because of a bug in FF. For outgoing simulcast I think we should make both 1 and 2 work correctly.

kc5nra commented:
I'll try to get to this soon - going on a family trip for 4 days. Thanks for the prompt attention and help :-)

kc5nra commented:
Hey @algesten, added a gentle attempt at an implementation to get a discussion started. I wasn't sure the best place to put the idea of forcing the ssrc-group creation, so I just went with one of the possible ways. (Versus a new argument to add_media, etc.)

In addition I added an optimization for mid/rid sending that I ran across in libwebrtc that seemed apropos

kc5nra commented on src/streams/send.rs:
```diff
@@ -382,8 +389,23 @@ impl StreamTx {
+
         // This is true also for RTX.
-        header_ref.ext_vals.mid = Some(mid);
+        if !recv_acked_ssrc {
```
Do we want to pull this from the stats with a slightly more opaque check? `if self.stats.rtt.is_some()` ? We can also pull this commit into its own PR

algesten replied:
This is good. I'm renaming it to `remote_acked_ssrc`

algesten commented:
Hi @kc5nra, I don't think we need a force_ssrc_group. If we just ensure we allocated SSRC for the simulast layers, we can always send the information as pre-communicated SSRC (case 1). Similarly case 2, we just need to double check we are setting the RTP headers, and always send it.

Whether this will trigger a bug in Firefox, we will see, but the ambition is directionally correct :)

kc5nra commented:
Sounds good - from a publisher to SFU (str0m not acting as the SFU but just the publisher), I was just trying to allow SDP generation of both variants (1) all the SSRCs enumerated + groups (2) just the simulcast and rids. This just allowed it to more closely resemble a libwebrtc offer. Not critical

algesten commented:
Just went down an SDP rabbit hole trying to reason about RID.

When the OFFER sends RIDs, I assume we must interpret the intention the [same as for PT](https://github.com/algesten/str0m/blob/main/SDP.md). I.e. if the direction is `sendrecv` or `recvonly`, the OFFER means: "I will only accept to _receive_ these RIDs". And if the direction is `sendonly`, it means "I propose to send these RIDs".

If this is consistent with PT logic, it means if the ANSWER contains different RIDs to the OFFER:

1. ANSWER with `sendrecv` similarly means _this is what I expect to receive_. This is asymetrical RIDs per direction. We might want to support it even if we don't support it for PT?
2. ANSWER with `sendonly` (mirroring `recvonly` OFFER), means, this is what I'm going to send, and it _MUST_ match the receiver (since the "I am prepared to receive" intention is the highest prio). This is definitely grounds for rejecting the ANSWER. There maybe a case for allowing a subset of the RIDs in the OFFER, but completely different is probably incorrect.
3. ANSWER with `recvonly` (mirroring `sendonly` OFFER), we must follow, the other side said how they _MUST_ receive. This allows the other side to remove RID they don't need/want.

It seems reasonable that a side can say "I don't want the high layer". Wonder if there are any specs around this. 

Bidirectional m-lines. The bane of my life. SDP sucks.

kc5nra commented:
This is great info, thanks - and your changes make a lot more sense.. I'm still getting a lay of the land!

8853 5.3.2

<snip>
An answerer that receives an offer with simulcast that lists a number of simulcast streams MAY reduce the number of simulcast streams in the answer, but it MUST NOT add simulcast streams.
</snip>

algesten commented:
> 8853 5.3.2
> An answerer that receives an offer with simulcast that lists a number of simulcast streams MAY reduce the number of simulcast streams in the answer, but it MUST NOT add simulcast streams.

Excellent!

Ok. So There are few things to do still:

1. Reject incorrect ANSWERs (i.e. that introduces new RIDs when they are not allowed to
2. Change `Writer` to only allow the RIDs that are int he new `Media::rids_tx`
3. Fix so DirectApi makes the correct changes to `rids_tx`

algesten commented:
Rebased off main to get rid of clippy errors.

algesten reviewed (approved):
I think we're ready to merge

kc5nra commented:
I'll run one round of testing in the morning if we are okay waiting a little bit - how do we want to squash these?

And thanks for the attentiveness in this :-)

algesten commented:
@kc5nra I just now squashed all the commits together to the two commits I think we should have.

Thanks for persevering through this!

 I'm fine with waiting to hit merge. Test away!

kc5nra commented:
@algesten I think we are good here, thanks!

algesten commented:
Excellent! Thanks for the help!
